language/stdlib/modules/compare.move

```diff
@@ -0,0 +1,99 @@
+// This defines a total order `Compare::cmp<T>` among Move values of the same type `T`. That is, we
+// have the guarantee:
+// `Compare::cmp<T>(t1, t2) == LESS_THEN` iff `Compare::cmp<T>(t2, t1) == GREATER_THEN`
+//
+// Move has bytecodes for polymorphic equality `==` and inequality `!=`. The implementation
+// of `Compare::cmp<T>(t1, t2)` also guarantees that
+// `Compare::cmp<T>(t1, t2) == EQUAL` iff `t1 == t2` and (similarly)
+// `Compare::cmp<T>(t1, t2) != EQUAL` iff `t1 != t2`.
+//
+// We implement `Compare.cmp(t1, t2)` using the `vector<u8>`'s produced by `LCS::to_bytes(t1)` and
+// `LCS::to_bytes(t2)`. We compare the vectors first by length, then byte-by-byte from right to left
+// if the lengths are equal. The length comparison is done in the obvious way; we return
+// `LESS_THAN` if len(lcs(t1)) < len(lcs(t2)) and `GREATER_THAN` if len(lcs(t1)) > len(lcs(t2)).
+// This ensures that for all primitive types `T` with `<` and `>` comparison operators exposed in
+// Move bytecode (`u8`, `u64`, `u128`), we have
+// `Compare::cmp<T>(t1, t2) == LESS_THEN` iff `t1 < t2` and (similarly)
+// `Compare::cmp<T>(t1, t2) == GREATER_THEN` iff `t1 > t2`.
+//
+// For all other types, the order is whatever the LCS encoding of the type and the comparison
+// strategy above gives you. One case where the order might be surprising is the `address` type.
+// Addresses are 16 byte hex values that LCS encodes with the identity function. The right to left,
+// byte-by-byte comparison means that (for example)
+// `Compare::cmp(0x01, 0x10)` == LESS_THAN` (as you'd expect), but
+// `Compare::cmp(0x100, 0x001) == LESS_THAN` (as you probably wouldn't expect). Keep this in mind
+// when writing code that compares addresses.
+//
+// Finally, a note on the constants `EQUAL`, `LESS_THAN`, and `GREATER_THAN`. Move does not have
+// signed integers, so we cannot use the usual 0, -1, 1 convention for this. Instead, we define a
+// new convention using u8 constants:
+// const EQUAL = 0u8
+// const LESS_THAN = 1u8
+// const GREATER_THAN = 2u8
+
+address 0x0:
+module Compare {
+    use 0x0::LCS;
+    use 0x0::Vector;
+
+    // Polymorphic comparison based on the LCS encoding of `v1` and `v2`. The lengths of the
+    // LCS-encoded vector are compared first, then the vectors are compared byte-by-byte from right
+    // to left if the lengths are equal.
+    // Returns either `EQUAL` (0u8), `LESS_THAN` (1u8), or `GREATER_THAN` (2u8).
+    public fun cmp<Value>(v1: &Value, v2: &Value): u8 {
```

Comment:
I'll take a deeper look tomorrow, but in the absence of enums (for now), thoughts on using a struct here?

You could return `Order`

And then add three functions 
```
public fun is_less(order: &Order): bool { .. }
public fun is_equal(order: &Order): bool { .. }
public fun is_greater(order: &Order): bool { .. }
```

I think it might be a bit easier to use than developing a new encoding of 0, 1, or 2

Reply:
- I'm not opposed to adding additional wrappers that do this, but I really want the basic API as well. The `Order struct is a bit heavier than a `u8`, and so is the additional call to `is_less` that you have to do after calling `cmp` (as opposed to just doing `== 0u8`  locally). Since the main use-case for this comparison is implementing collections (I don't expect a lot of user code to call it), I don't want to be forced to pay for that overhead.
- Once we have constants, I think this code/conveniton will be much more readable.
- I don't think adding additional types is always more useable, particularly if you have to do extra function calls to interpret them. @runtian-zhou and I went through a variant of this with the `Time` API where we went through several iterations of different type-based representations before eventually settling on a plain u64 (which ended  up making  the other code a lot more readable). This code is a different beast of course; just pointing this out (as someone who actually likes mistake-preventing type wrappers a lot).

Reply:
These seem like good answers. While I'd still prefer enums, constants here seem like a good enough developer experience